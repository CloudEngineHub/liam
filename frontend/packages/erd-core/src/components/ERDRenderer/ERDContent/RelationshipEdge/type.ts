import type { Cardinality } from '@liam-hq/db-structure'
import type { Edge } from '@xyflow/react'

export type Data = {
  isHighlighted: boolean
  cardinality: Cardinality | undefined
}

export type RelationshipEdgeType = Edge<Data, 'relationship'>
