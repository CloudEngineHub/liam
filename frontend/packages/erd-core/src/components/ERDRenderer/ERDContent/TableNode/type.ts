import type { Cardinality, Table } from '@liam-hq/db-structure'
import type { Node } from '@xyflow/react'

export type Data = {
  table: Table
  isActiveHighlighted: boolean
  isHighlighted: boolean
  highlightedHandles: string[]
  sourceColumnName: string | undefined
  targetColumnCardinalities?:
    | Record<string, Cardinality | undefined>
    | undefined
}

export type TableNodeType = Node<Data, 'table'>
