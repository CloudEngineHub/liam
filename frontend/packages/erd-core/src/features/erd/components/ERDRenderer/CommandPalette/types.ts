export type CommandPaletteInputMode = { type: 'default' } | { type: 'command' }
// upcoming input mode
// | { type: 'column'; tableName: string }
