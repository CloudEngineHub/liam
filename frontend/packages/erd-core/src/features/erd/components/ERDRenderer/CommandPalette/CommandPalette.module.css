.content[cmdk-dialog] {
  position: fixed;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  width: 638px;
  height: 508px;
  border-radius: var(--border-radius-lg);
  box-shadow: 0px 4px 20px 0px var(--shadow-basic-shadow, #000);
  animation: contentShow 150ms cubic-bezier(0.16, 1, 0.3, 1);
  background-color: var(--pane-background);
  color: var(--global-foreground);
  border: 1px solid var(--global-border);
  z-index: 10000;
}

[cmdk-overlay]:has(+ .content) {
  position: fixed;
  inset: 0;
  width: 100vw;
  height: 100vh;
  backdrop-filter: blur(8px);
  animation: overlayShow 150ms cubic-bezier(0.16, 1, 0.3, 1);
  z-index: 10000;
}

@keyframes overlayShow {
  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
}

@keyframes contentShow {
  from {
    opacity: 0;
    transform: translate(-50%, -48%) scale(0.96);
  }

  to {
    opacity: 1;
    transform: translate(-50%, -50%) scale(1);
  }
}
