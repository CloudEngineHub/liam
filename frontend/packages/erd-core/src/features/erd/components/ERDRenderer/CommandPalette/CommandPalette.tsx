'use client'

import { DialogDescription, DialogTitle } from '@radix-ui/react-dialog'
import { Command } from 'cmdk'
import { type FC, useEffect, useState } from 'react'
import styles from './CommandPalette.module.css'

export const CommandPalette: FC = () => {
  const [open, setOpen] = useState(false)

  // Toggle the menu when âŒ˜K is pressed
  useEffect(() => {
    const down = (event: KeyboardEvent) => {
      if (event.key === 'k' && (event.metaKey || event.ctrlKey)) {
        event.preventDefault()
        setOpen((open) => !open)
      }
    }

    document.addEventListener('keydown', down)
    return () => document.removeEventListener('keydown', down)
  }, [])

  return (
    <Command.Dialog
      open={open}
      onOpenChange={setOpen}
      contentClassName={styles.content}
    >
      <DialogTitle hidden>Command Palette</DialogTitle>
      <DialogDescription hidden>
        A search-based interface that allows quick access to various commands
        and features within the application.
      </DialogDescription>
      Coming Soon...
    </Command.Dialog>
  )
}
