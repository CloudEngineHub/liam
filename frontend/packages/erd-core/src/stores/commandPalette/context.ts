import { createContext } from 'react'

type CommandPaletteContextValue = {
  open: boolean
  setOpen: (open: boolean) => void
}

export const CommandPaletteContext =
  createContext<CommandPaletteContextValue | null>(null)
