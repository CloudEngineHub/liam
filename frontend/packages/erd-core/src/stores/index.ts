export * from './dbStructure'
export * from './userEditing'
