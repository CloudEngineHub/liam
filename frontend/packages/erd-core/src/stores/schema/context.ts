import { schemaDiffItemsSchema, schemaSchema } from '@liam-hq/db-structure'
import { createContext } from 'react'
import * as v from 'valibot'

const schemaStoreSchema = v.object({
  current: schemaSchema,
  previous: v.optional(schemaSchema),
  diffItems: v.optional(schemaDiffItemsSchema),
})

export type SchemaContextValue = v.InferOutput<typeof schemaStoreSchema>

export const SchemaContext = createContext<SchemaContextValue | null>(null)
