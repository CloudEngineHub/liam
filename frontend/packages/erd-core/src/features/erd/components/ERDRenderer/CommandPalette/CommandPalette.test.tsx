import { aTable } from '@liam-hq/db-structure'
import { cleanup, render, screen, within } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { ReactFlowProvider } from '@xyflow/react'
import { NuqsTestingAdapter } from 'nuqs/adapters/testing'
import { type FC, type ReactNode, useContext } from 'react'
import { afterEach, describe, expect, it } from 'vitest'
import { SchemaProvider, UserEditingProvider } from '@/stores'
import type { SchemaStore } from '@/stores/schema/schema'
import { UserEditingContext } from '@/stores/userEditing/context'
import { CommandPalette } from './CommandPalette'

afterEach(() => {
  cleanup()
})

const schema: SchemaStore = {
  current: {
    tables: {
      users: aTable({ name: 'users' }),
      posts: aTable({ name: 'posts' }),
      follows: aTable({ name: 'follows' }),
      user_settings: aTable({ name: 'user_settings' }),
    },
    tableGroups: {},
    relationships: {},
  },
}

const ActiveTableNameDisplay: FC = () => {
  const userEditing = useContext(UserEditingContext)

  return (
    //  The currently selected table name is displayed via Context, and used in tests for assertions.
    <div data-testid="test-active-table-name-display">
      {userEditing?.activeTableName}
    </div>
  )
}

const wrapper = ({ children }: { children: ReactNode }) => (
  <NuqsTestingAdapter>
    <ReactFlowProvider>
      <UserEditingProvider>
        <ActiveTableNameDisplay />
        <SchemaProvider schema={schema}>{children}</SchemaProvider>
      </UserEditingProvider>
    </ReactFlowProvider>
  </NuqsTestingAdapter>
)

const prepareCommandPalette = async () => {
  const user = userEvent.setup()

  render(<CommandPalette />, { wrapper })

  await user.keyboard('{Meta>}k{/Meta}')
  const dialog = await screen.findByRole('dialog', {
    name: 'Command Palette',
  })
  const searchCombobox = within(dialog).getByRole('combobox')
  const preview = within(dialog).getByTestId('CommandPalettePreview')

  return { user, elements: { dialog, searchCombobox, preview } }
}

it('displays nothing by default', () => {
  render(<CommandPalette />, { wrapper })

  expect(screen.queryByRole('dialog')).not.toBeInTheDocument()
})

describe('dialog opening interactions', () => {
  it('opens the dialog with typing âŒ˜K', async () => {
    const user = userEvent.setup()

    render(<CommandPalette />, { wrapper })

    await user.keyboard('{Meta>}k{/Meta}')

    expect(
      screen.getByRole('dialog', { name: 'Command Palette' }),
    ).toBeInTheDocument()
  })

  it('opens the dialog with typing ctrl+K', async () => {
    const user = userEvent.setup()

    render(<CommandPalette />, { wrapper })

    await user.keyboard('{Control>}k{/Control}')

    expect(
      screen.getByRole('dialog', { name: 'Command Palette' }),
    ).toBeInTheDocument()
  })
})

describe('options and combobox interactions', () => {
  it('renders options with table name', async () => {
    const {
      elements: { dialog },
    } = await prepareCommandPalette()

    expect(within(dialog).getAllByRole('option')).toHaveLength(4)
    expect(within(dialog).getByRole('option', { name: 'users' }))
    expect(within(dialog).getByRole('option', { name: 'posts' }))
    expect(within(dialog).getByRole('option', { name: 'follows' }))
    expect(within(dialog).getByRole('option', { name: 'user_settings' }))
  })

  it('filters options based on user input in the combobox', async () => {
    const {
      user,
      elements: { dialog, searchCombobox },
    } = await prepareCommandPalette()

    expect(searchCombobox).toHaveFocus()

    await user.keyboard('user')

    expect(within(dialog).getAllByRole('option')).toHaveLength(2)
    expect(within(dialog).getByRole('option', { name: 'users' }))
    expect(within(dialog).getByRole('option', { name: 'user_settings' }))
  })

  it('renders "No results found." if user input does not match any options', async () => {
    const {
      user,
      elements: { dialog, searchCombobox },
    } = await prepareCommandPalette()

    expect(searchCombobox).toHaveFocus()

    await user.keyboard('HelloWorld')

    expect(within(dialog).queryByRole('option')).not.toBeInTheDocument()
    expect(within(dialog).getByText('No results found.')).toBeInTheDocument()
  })
})

describe('preview with option interactions', () => {
  it('displays a preview of the option hovered', async () => {
    const {
      user,
      elements: { dialog, preview },
    } = await prepareCommandPalette()

    expect(within(preview).getByText('users')).toBeInTheDocument()

    await user.hover(within(dialog).getByRole('option', { name: 'follows' }))

    expect(within(preview).getByText('follows')).toBeInTheDocument()
  })

  it('displays a preview of the option selected via arrow key navigation', async () => {
    const {
      user,
      elements: { preview },
    } = await prepareCommandPalette()

    expect(within(preview).getByText('users')).toBeInTheDocument()

    await user.keyboard('{ArrowDown}')
    expect(within(preview).getByText('posts')).toBeInTheDocument()

    await user.keyboard('{ArrowDown}')
    expect(within(preview).getByText('follows')).toBeInTheDocument()

    await user.keyboard('{ArrowUp}')
    expect(within(preview).getByText('posts')).toBeInTheDocument()
  })
})

describe('go to ERD with option select', () => {
  it('go to the table of clicked option and close dialog', async () => {
    const {
      user,
      elements: { dialog },
    } = await prepareCommandPalette()

    expect(
      screen.getByTestId('test-active-table-name-display'),
    ).toBeEmptyDOMElement()

    await user.click(within(dialog).getByRole('option', { name: 'follows' }))

    expect(
      screen.getByTestId('test-active-table-name-display'),
    ).toHaveTextContent('follows')
    expect(dialog).not.toBeInTheDocument()
  })

  it('go to the table of selected option by typing Enter key and close dialog', async () => {
    const {
      user,
      elements: { dialog, preview },
    } = await prepareCommandPalette()

    expect(
      screen.getByTestId('test-active-table-name-display'),
    ).toBeEmptyDOMElement()

    // select "posts" option by typing Enter key
    await user.keyboard('{ArrowDown}')
    expect(within(preview).getByText('posts')).toBeInTheDocument()
    await user.keyboard('{Enter}')

    expect(dialog).not.toBeInTheDocument()
    expect(
      screen.getByTestId('test-active-table-name-display'),
    ).toHaveTextContent('posts')
  })
})
