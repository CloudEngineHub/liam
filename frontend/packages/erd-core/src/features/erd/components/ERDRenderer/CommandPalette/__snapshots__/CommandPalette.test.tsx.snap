// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`snapshot test 1`] = `
<div
  aria-describedby="radix-«r51»"
  aria-labelledby="radix-«r50»"
  class="_content_933568"
  data-state="open"
  id="radix-«r4v»"
  role="dialog"
  style="pointer-events: auto;"
  tabindex="-1"
>
  <h2
    hidden=""
    id="radix-«r50»"
  >
    Command Palette
  </h2>
  <p
    hidden=""
    id="radix-«r51»"
  >
    A search-based interface that allows quick access to various commands and features within the application.
  </p>
  <div
    cmdk-root=""
    tabindex="-1"
  >
    <label
      cmdk-label=""
      for="radix-«r54»"
      id="radix-«r53»"
      style="position: absolute; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0, 0, 0, 0); white-space: nowrap; border-width: 0px;"
    />
    <div
      class="_searchContainer_6f7bc2"
    >
      <div
        class="_searchFormWithIcon_6f7bc2"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-search _searchIcon_6f7bc2"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m21 21-4.34-4.34"
          />
          <circle
            cx="11"
            cy="11"
            r="8"
          />
        </svg>
        <input
          aria-autocomplete="list"
          aria-controls="radix-«r52»"
          aria-expanded="true"
          aria-labelledby="radix-«r53»"
          autocomplete="off"
          autocorrect="off"
          cmdk-input=""
          id="radix-«r54»"
          placeholder="Search"
          role="combobox"
          spellcheck="false"
          type="text"
          value=""
        />
      </div>
      <button
        class="_wrapper_ba280b _xs_ba280b _escButton_6f7bc2 _outlineSecondary_ba280b"
        type="button"
      >
        ESC
      </button>
    </div>
    <div
      class="_main_6f7bc2"
    >
      <div
        aria-label="Suggestions"
        cmdk-list=""
        id="radix-«r52»"
        role="listbox"
        tabindex="-1"
      >
        <div
          cmdk-list-sizer=""
        >
          <div
            cmdk-group=""
            data-value="Tables"
            role="presentation"
          >
            <div
              aria-hidden="true"
              cmdk-group-heading=""
              id="radix-«r56»"
            >
              Tables
            </div>
            <div
              aria-labelledby="radix-«r56»"
              cmdk-group-items=""
              role="group"
            >
              <a
                aria-disabled="false"
                aria-selected="true"
                cmdk-item=""
                data-disabled="false"
                data-selected="true"
                data-value="users"
                href="?active=users"
                id="radix-«r57»"
                role="option"
              >
                <svg
                  aria-hidden="true"
                  class="lucide lucide-table2 lucide-table-2 _itemIcon_6f7bc2"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M9 3H5a2 2 0 0 0-2 2v4m6-6h10a2 2 0 0 1 2 2v4M9 3v18m0 0h10a2 2 0 0 0 2-2V9M9 21H5a2 2 0 0 1-2-2V9m0 0h18"
                  />
                </svg>
                <span
                  class="_itemText_6f7bc2"
                >
                  users
                </span>
              </a>
              <a
                aria-disabled="false"
                aria-selected="false"
                cmdk-item=""
                data-disabled="false"
                data-selected="false"
                data-value="posts"
                href="?active=posts"
                id="radix-«r58»"
                role="option"
              >
                <svg
                  aria-hidden="true"
                  class="lucide lucide-table2 lucide-table-2 _itemIcon_6f7bc2"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M9 3H5a2 2 0 0 0-2 2v4m6-6h10a2 2 0 0 1 2 2v4M9 3v18m0 0h10a2 2 0 0 0 2-2V9M9 21H5a2 2 0 0 1-2-2V9m0 0h18"
                  />
                </svg>
                <span
                  class="_itemText_6f7bc2"
                >
                  posts
                </span>
              </a>
              <a
                aria-disabled="false"
                aria-selected="false"
                cmdk-item=""
                data-disabled="false"
                data-selected="false"
                data-value="follows"
                href="?active=follows"
                id="radix-«r59»"
                role="option"
              >
                <svg
                  aria-hidden="true"
                  class="lucide lucide-table2 lucide-table-2 _itemIcon_6f7bc2"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M9 3H5a2 2 0 0 0-2 2v4m6-6h10a2 2 0 0 1 2 2v4M9 3v18m0 0h10a2 2 0 0 0 2-2V9M9 21H5a2 2 0 0 1-2-2V9m0 0h18"
                  />
                </svg>
                <span
                  class="_itemText_6f7bc2"
                >
                  follows
                </span>
              </a>
              <a
                aria-disabled="false"
                aria-selected="false"
                cmdk-item=""
                data-disabled="false"
                data-selected="false"
                data-value="user_settings"
                href="?active=user_settings"
                id="radix-«r5a»"
                role="option"
              >
                <svg
                  aria-hidden="true"
                  class="lucide lucide-table2 lucide-table-2 _itemIcon_6f7bc2"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M9 3H5a2 2 0 0 0-2 2v4m6-6h10a2 2 0 0 1 2 2v4M9 3v18m0 0h10a2 2 0 0 0 2-2V9M9 21H5a2 2 0 0 1-2-2V9m0 0h18"
                  />
                </svg>
                <span
                  class="_itemText_6f7bc2"
                >
                  user_settings
                </span>
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="_previewContainer_6f7bc2"
        data-testid="CommandPalettePreview"
      >
        <div
          class="_previewBackground_6f7bc2"
        >
          <div
            class="_wrapper_f5909b"
            data-erd="false"
            data-state="closed"
          >
            <div
              class="_wrapper_5b50e5"
            >
              <div
                class="_container_5b50e5"
              >
                <svg
                  aria-hidden="true"
                  class="lucide lucide-table2 lucide-table-2 _tableIcon_5b50e5"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M9 3H5a2 2 0 0 0-2 2v4m6-6h10a2 2 0 0 1 2 2v4M9 3v18m0 0h10a2 2 0 0 0 2-2V9M9 21H5a2 2 0 0 1-2-2V9m0 0h18"
                  />
                </svg>
                <span
                  class="_name_5b50e5"
                >
                  users
                </span>
              </div>
            </div>
            <ul />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
